File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        stack <pair <int, int>> st;
        heights.push_back(0);
        int maxArea = 0;
        for (int i = 0; i < heights.size(); i++) {
            int start = i;
            while (!st.empty() && st.top().second > heights[i]) {
                auto [idx, height] = st.top();
                st.pop();
                maxArea = max(maxArea, height * (i - idx));
                start = idx;
            }
            st.push({start, heights[i]});
        }
        return maxArea;
    }
};

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, h: List[int]) -> int:
        stack = []
        max_area = 0
        h.append(0)

        for idx, height in enumerate(h):
            start_idx = idx
            while stack and stack[-1][1] > height:
                curr_idx, curr_height = stack.pop()
                max_area = max(max_area, curr_height * (idx - curr_idx))
                start_idx = curr_idx
            stack.append((start_idx, height))

        # for idx, height in stack:
        #     max_area = max(max_area, height * (len(h) - idx))

        return max_area